src/Servers/Kestrel/Core/src/KestrelConfigurationLoader.cs

```diff
@@ -435,14 +436,98 @@ private X509Certificate2 LoadCertificate(CertificateConfig certInfo, string endp
             }
             else if (certInfo.IsFileCert)
             {
-                var env = Options.ApplicationServices.GetRequiredService<IHostEnvironment>();
-                return new X509Certificate2(Path.Combine(env.ContentRootPath, certInfo.Path), certInfo.Password);
+                var environment = Options.ApplicationServices.GetRequiredService<IHostEnvironment>();
+                if (certInfo.KeyPath != null)
+                {
+                    X509Certificate2 certificate = null;
+                    if (TryReadPemRSAKey(certInfo, environment, out var rsaKey))
+                    {
+                        var publicCertificate = new X509Certificate2(Path.Combine(environment.ContentRootPath, certInfo.Path));
+                        certificate =  publicCertificate.CopyWithPrivateKey(rsaKey);
+                    }
+
+                    if (TryReadPemDSAKey(certInfo, environment, out var dsaKey))
+                    {
+                        var publicCertificate = new X509Certificate2(Path.Combine(environment.ContentRootPath, certInfo.Path));
+                        certificate = publicCertificate.CopyWithPrivateKey(dsaKey);
+                    }
+
+                    if (certificate != null)
+                    {
+                        if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
+                        {
+                            return PersistKey(certificate);
+                        }
+
+                        return certificate;
+                    }
+
+                    throw new InvalidOperationException(CoreStrings.InvalidPemKey);
+                }
+
+                return new X509Certificate2(Path.Combine(environment.ContentRootPath, certInfo.Path), certInfo.Password);
             }
             else if (certInfo.IsStoreCert)
             {
                 return LoadFromStoreCert(certInfo);
             }
             return null;
+
+            static X509Certificate2 PersistKey(X509Certificate2 fullCertificate)
+            {
+                // We need to force the key to be persisted.
```

Comment:
Persisted to where?

Reply:
On Windows this needs to be super careful, otherwise you can end up filling the private key directory. @bartonjs will know more

Reply:
This is actually doing what I call a "perphemeral" key, in that it's "temporarily persisted", as long as the cert gets freed (Dispose or Finalize) the key-on-disk gets erased.

Unfortunately, SChannel doesn't work without a key-on-disk (or key-in-HSM), so this _is_ required after loading the private key from a PEM file.

Reply:
> as long as the cert gets freed (Dispose or Finalize) the key-on-disk gets erased.

aspnet people, is the certificate automatically disposed when the host is shutdown?

A test that verifies that PEM keys are cleaned up on shutdown would be a good idea.

Reply:
> A test that verifies that PEM keys are cleaned up on shutdown would be a good idea.

Well.  Sort of.  That's a really hard test to write.  It requires plumbing in a lot of knowledge of Windows Cryptography that even the crypto layer doesn't use automated tests for, since it's potentially subject to change.

Reply:
> aspnet people, is the certificate automatically disposed when the host is shutdown?

No it's not. We don't call dispose anywhere on the cert provided to Kestrel AFAIK

Reply:
But it would get finalized.

Reply:
.NET Core doesn't guarantee finalizers on exit (IIRC).  Unless the host (Kestrel?) calls GC.Collect(); GC.WaitForPendingFinalizers(); on a normal shutdown.

Reply:
To be fair, the behavior of having used this method is the same as whatever happens when someone specifies a PFX.  Or loads a PFX themselves and passes in an object (if that's a thing).  Since, well, it's just loading a PFX :smile:.

Reply:
If an app is shutting down (e.g. ctrl+c was pressed), would the process wait around to run finalizers?

I think we should track "temporarily persisted" X509 certs and ensure they are disposed.

Reply:
> To be fair, the behavior of having used this method is the same as whatever happens when someone specifies a PFX. Or loads a PFX themselves and passes in an object (if that's a thing). Since, well, it's just loading a PFX 😄.

Yep this was my reasoning in doing so.

Reply:
> I think we should track "temporarily persisted" X509 certs and ensure they are disposed.

We haven't done this since we added support for config (we should track regular pfx files too if that's the case) and this hasn't been an issue in practice. While I agree that the issue is real, I'm not sure it is something that happens regularly in practice.

I'm not sure if this is because it is hard to reach the upper limit of keys in the underlying abstraction or if it is because people is very diligent in disposing their certificates. I'm siding more with 1. I suggest that we place tracking and disposal of loaded certificates as a separate item.

WRT to loading the keys, the code in here as @bartonjs mentions is equivalent to what happens when you create a new `X509Certiificate2` instance by hand from a pfx.

Reply:
I've filed https://github.com/dotnet/aspnetcore/issues/23622 for tracking disposing the certs on graceful shutdown.